olm/operator-registry.md

```diff
@@ -0,0 +1,291 @@
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+    
+`operator-registry add --batch`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without $operatorName included
+    ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without latest version of $operatorName included
+    ex: `operator-registry delete-latest foo example.db`
+    
+*Reference non latest versions of bundles by image digests*
+
+Currently the operator-registry database has a table to the effect of:
+
+OperatorBundle
+    -name: foo
+    -bundle: "{$jsonblob}"
+    
+We will add a field to this table that will reference the bundle image path and, when the operator is not the latest version of the default channel, will not include the bundle blob itself:
+
+OperatorBundle
+    -name: foo
+    -bundle: NULL
+    -bundlePath: "quay.io/community-operators/foo:sha256@abcd123"
+    
+Additionally, we will update the grpc API that OLM currently uses to get the bundle that if it attempts to pull the non latest version will return null. In that case, it will make a second query to get the bundlePath. Then we will create a pod with that image that will serve the bundle data by writing the data to a configmap that can be read by OLM.
+
+In order to serve that manifest image, we cannot pull the image directly. Instead we will:
+
+1. Create a pod spec that uses an init container to inject a binary into the manifest image.
+2. Run the manifest image with that binary that writes to a configmap in a well known location
+3. Read the configmap data
+4. On return kill the pod
+5. Delete the configmap
```

Comment:
Are you intending to start backing the operator registry pods containing the database with actual persisted storage? If not, and the operator registry pod is killed for whatever reason, then you've lost all of those modifications you made to the index and would have to re-run any of these pods that had been launched. Leaving the ConfigMap in place instead of deleting it would provide persistence for that data if the operator registry gets restarted and needs to be rebuilt.

Leaving the ConfigMaps does mean more management later, need to know when to clean them up (like pruning old versions that are no longer needed).

Reply:
Actually, we are not going to update the database at all at runtime. This implementation means that operator-registries are still immutable once built. Right now this bundle request is only used to explicitly create the CSV resource on the cluster -- we will just use that metadata to persist the data on that object.

It is true that if the CSV is explicitly deleted for some reason (most likely install) that this process would happen again. But at the moment the frequency of that is so low that trying to persist these configmaps and handle their state is probably not worth that optimization. Obviously though if we decide to remove the CSV API altogether we may have to revisit this (mutable database, persist them as config maps, etc)

Reply:
CSV does not currently contain all the data needed to install for some operators, and I don't think that it makes sense to assume that. I would like to think about if we need to cache this, or if re-pulling/running this object is acceptable. 

Are there times when we need the old files (CRD's maybe?) for upgrade? Imagine the skip_range case where this old bundle ref may be removed from the manifest-image.

Reply:
>  I would like to think about if we need to cache this, or if re-pulling/running this object is acceptable.

This enhancement does not propose removing InstallPlans, which perform the caching/recreation you're concerned with. When/if we remove them, we can discuss alternative caching solutions.